service/worker/sysworkflow/system_workflow.go

```diff
@@ -22,101 +22,341 @@ package sysworkflow
+			return errArchivalUploadActivityMarshalBlob
+		}
+		tags, err := ConvertHeaderToTags(historyBlob.Header)
+		if err != nil {
+			logger.WithError(err).Error("failed to convert header to tags, stopping archival upload")
+			metricsClient.IncCounter(metrics.ArchivalUploadActivityScope, metrics.SysWorkerConvertHeaderToTagsFailures)
+			return errArchivalUploadActivityConvertHeaderToTags
+		}
+		wrapFunctions := []blob.WrapFn{blob.JSONEncoded()}
+		if container.Config.EnableArchivalCompression(domainCacheEntry.GetInfo().Name) {
+			wrapFunctions = append(wrapFunctions, blob.GzipCompressed())
+		}
+		currBlob, err := blob.Wrap(blob.NewBlob(body, tags), wrapFunctions...)
+		if err != nil {
+			logger.WithError(err).Error("failed to wrap blob, stopping archival upload")
+			metricsClient.IncCounter(metrics.ArchivalUploadActivityScope, metrics.SysWorkerWrapBlobFailures)
+			return errArchivalUploadActivityWrapBlob
+		}
+		if err := blobUploadRetryForever(blobstoreClient, bucket, key, currBlob); err != nil {
+			logger.WithError(err).Error("failed to upload blob, stopping archival upload")
+			metricsClient.IncCounter(metrics.ArchivalUploadActivityScope, metrics.SysWorkerBlobUploadNonRetryableFailures)
+			return errArchivalUploadActivityUploadBlob
+		}
+		activity.RecordHeartbeat(ctx, historyBlob.Header.CurrentPageToken)
+	}
 	return nil
 }
 
+// ArchivalDeleteHistoryActivity deletes the workflow execution history from persistence.
+// All retryable errors are retried forever. If an error is returned it is of type archivalActivityNonRetryableErr.
+func ArchivalDeleteHistoryActivity(ctx context.Context, request ArchiveRequest) error {
+	container := ctx.Value(sysWorkerContainerKey).(*SysWorkerContainer)
+	logger := container.Logger.WithFields(bark.Fields{
+		logging.TagArchiveRequestDomainID:          request.DomainID,
+		logging.TagArchiveRequestWorkflowID:        request.WorkflowID,
+		logging.TagArchiveRequestRunID:             request.RunID,
+		logging.TagArchiveRequestEventStoreVersion: request.EventStoreVersion,
+		logging.TagArchiveRequestLastFirstEventID:  request.LastFirstEventID,
+	})
+	metricsClient := container.MetricsClient
+	if request.EventStoreVersion == persistence.EventStoreVersionV2 {
+		err := persistence.DeleteWorkflowExecutionHistoryV2(container.HistoryV2Manager, request.BranchToken, container.Logger)
+		if err == nil {
+			return nil
+		}
+		op := func() error {
+			return persistence.DeleteWorkflowExecutionHistoryV2(container.HistoryV2Manager, request.BranchToken, container.Logger)
+		}
+		for err != nil && common.IsPersistenceTransientError(err) {
+			err = backoff.Retry(op, common.CreatePersistanceRetryPolicy(), common.IsPersistenceTransientError)
+		}
+		logger.WithError(err).Error("failed to delete history from events v2")
+		metricsClient.IncCounter(metrics.ArchivalDeleteHistoryActivityScope, metrics.SysWorkerDeleteHistoryV2NonRetryableFailures)
+		return errDeleteHistoryActivityDeleteFromV2
```

Comment:
- i believe this error is declared as non-retryable in the activity retry policy options. However, your backoff.Retry() will give up all retries in about 30s. Isn't this the only place where history gets deleted ? If yes, you should retry infinitely

- activity options has heartbeat timeout set to 10s - if you are midst of a retry loop, you want to avoid timing out on heartbeat

Reply:
- I am retrying indefinitely. The backoff.Retry() is inside of a loop that continues as long as error type is transient. 
- On the heartbeat time out issue. I see what you mean. I think it's actually related to how we decide to handle heartbeats more generally.